### PR TITLE
sinkv2(cdc): add interface EventSortEngine

### PR DESCRIPTION
**Pull request by hicqu**
Signed-off-by: qupeng <qupeng@pingcap.com>

### What problem does this PR solve?

Issue Number: ref #5928 

### What is changed and how it works?

Add a new interface named EventSortEngine.

#### Tests 

 - No code

### Release note 

```release-note
None
```

### CODE REVIEW

File: pkg/sorter/sorter.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+// NOTE: All interfaces are thread-safe.
+type EventSortEngine interface {
+	// IsTableBased tells whether the sort engine is based on table or not.
+	// If it's based on table, fetching events by table is prefered.
+	IsTableBased() bool
+
+	// AddTable adds the table into the engine.
+	AddTable(tableID model.TableID)
+
+	// Remove removes the table from the engine.
```
Comment (hi-rustin) [inline pkg/sorter/sorter.go:31]:
```suggestion
	// RemoveTable removes the table from the engine.
```

File: pkg/sorter/sorter.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+	Next() (event *model.PolymorphicEvent, txnFinished Position, err error)
+
+	// Close closes the iterator.
+	Close() error
+}
+
+// Position is used to
+//  1. fetch or clear events from an engine, for example, see EventSortEngine.FetchByTable.
+//  2. calculate the next position with method Next.
+type Position struct {
```
Comment (hi-rustin) [inline pkg/sorter/sorter.go:95]: Could you please add a `Less` function to it?

**Review by hi-rustin — COMMENTED** [review]
rest LGTM

File: pkg/sorter/sorter.go (hunk 2 above)
Comment (hicqu) [reply pkg/sorter/sorter.go:95]: I prefer `Compare` instead of `Less`.

**Review by hi-rustin — APPROVED** [review]

File: pkg/sorter/sorter.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+	// lowerBound is inclusive and only resolved events can be retrieved.
+	//
+	// NOTE: FetchByTable is always available even if IsTableBased returns false.
+	FetchByTable(tableID model.TableID, lowerBound, upperBound Position) EventIterator
+
+	// FetchAllTables creates an iterator to fetch events from all tables.
+	// lowerBound is inclusive and only resolved events can be retrieved.
+	//
+	// NOTE: It's only available if IsTableBased returns true.
+	FetchAllTables(lowerBound Position) EventIterator
```
Comment (nongfushanquan) [inline pkg/sorter/sorter.go:57]: IsTableBased = false ？

Comment (hicqu) [reply pkg/sorter/sorter.go:57]: If it's table based, we can't fetch events from all tables. We can only fetch events with `FetchByTable`.

**Review by nongfushanquan — APPROVED** [review]

File: pkg/sorter/sorter.go (hunk 3 above)
Comment (hicqu) [reply pkg/sorter/sorter.go:57]: Yes you are right. It's a mistake.

**Pull request merged**